lib/chipmunk/package_storage.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+module Chipmunk
+  # PackageStorage is a factory for proxies to a Package's underlying storage.
+  # An instance supports a named set of format types, mapping each to a storage
+  # proxy class and creating an appropriate proxy instance for a package.
+  class PackageStorage
+
+    # Create a PackageStorage instance.
+    #
+    # @param volumes [Array<Chipmunk::Volume>] the storage volumes where packages are stored
+    def initialize(volumes:)
+      @volumes = Hash[volumes.map {|volume| [volume.name, volume] }]
+    end
+
+    # Retrieve the storage proxy for a given package. The package must be stored and of a registered format.
+    # @param package [Package] the package for which to retrieve a storage proxy
+    # @raise [Chipmunk::PackageNotStoredError] if the package is not stored
+    # @raise [Chipmunk::UnsupportedFormatError] if the package format is not registered
+    def for(package)
+      raise Chipmunk::PackageNotStoredError, package unless package.stored?
+
+      volume_for(package).get(package.storage_path)
+    end
+
+    # Move the source archive into preservation storage and update the package's
+    # storage_volume and storage_path accordingly.
+    def write(package, source)
+      if package.format == Chipmunk::Bag.format
+        move_bag(package, source)
+      else
+        raise Chipmunk::UnsupportedFormatError, "Package #{package.bag_id} has invalid format: #{package.format}"
+      end
+    end
+
+    private
+
+    def move_bag(package, source)
+      bag_id = package.bag_id
+      prefixes = bag_id.match(/^(..)(..)(..).*/)
+      raise "bag_id too short: #{bag_id}" unless prefixes
+
+      storage_path = File.join("/", prefixes[1..3], bag_id)
```

Comment:
The conversion of the storage path definitely seems like it belongs in `Volume` rather than here.

Reply:
The gotcha is that this path pattern is specific to storage in our native layout. Volume is context independent for now (incoming path strategy or storage path strategy is external). It may be that this should be pushed down into subclasses of Volume or some other, but I stopped at the first place I could behind the clean API (IncomingStorage or PackageStorage), so we can decide that later as a refactoring.